otherlibs/stdune/src/path.ml

```diff
@@ -1220,7 +1208,68 @@ let source s = in_source_tree s
 
 let build s = in_build_dir s
 
-module Table = Hashtbl.Make (T)
+module Table = struct
+  type 'a t =
+    { source : 'a Source0.Table.t
+    ; build : 'a Build.Table.t
+    ; external_ : 'a External.Table.t
+    }
+
+  let create size =
+    { source = Source0.Table.create size
```

Comment:
What's the initial size of Dune tables? Shall we do something here to avoid grabbing 3x size now?

Reply:
We could, but I actually think it doesn't matter. We only use hash tables for global caches and we really underestimate the sizes of those. I could divide each table size by 3 if you prefer though

Reply:
Or, maybe, use size 0 or make these tables lazy? I'm just observing some not very consistent benchmark results and wonder if this could be an explanation. Like, what if the majority of tables we have actually use just one of the path types?

Reply:
> I'm just observing some not very consistent benchmark results and wonder if this could be an explanation. 

Wait, I just realized that we barely use this table at all in memo. Instead, we instantiate the slow tables when we create the store 😄 

So this change does very very little in fact :)

> Like, what if the majority of tables we have actually use just one of the path types?

Don't think it would matter. The other tables would stay near their (tiny) initial sizes and the empty space in the one large table would easily dwarf the empty space in the other two.

Reply:
> The other tables would stay near their (tiny) initial sizes

Wouldn't it be better for them to stay the tiniest possible size of 0 rather than whatever `Table.t` happens to pick?

Reply:
> Wait, I just realized that we barely use this table at all in memo. Instead, we instantiate the slow tables when we create the store

Hmm, I guess we'd like to fix that? Not sure in which order (or together)?

Reply:
> Wouldn't it be better for them to stay the tiniest possible size of 0 rather than whatever Table.t happens to pick?

If that's the case, the caller can just pass 0 already. Perhaps we should just do that and get rid of this size argument? I don't particularly see a need for it anyway.

> Hmm, I guess we'd like to fix that? Not sure in which order (or together)?

If you confirm there's no regression from this PR, I think it's fine doing it after.

Reply:
> If you confirm there's no regression from this PR, I think it's fine doing it after.

The benchmarks I run are kind of all over the place memory-wise. And, weirdly, there seem to be a consistent small (0.3-1.0%) wall clock build time regression. I can try to rerun it tomorrow but it feels like something is wrong.

Reply:
I see. Well that's quite strange because the benchmark here shows a tiny improvement (within the noise though).

What I can also try is to add inlining annotations. I suppose it can hurt performance if there's an extra function call.

Reply:
It seemed to have make a difference for null builds (our only stable measurements for such small differences).

Reply:
Nice, I'll rerun the benchmarks.